### PR TITLE
UI infrastructure details

### LINKED ISSUE #41: Improve layout of infrastructure details page in UI
Some modifications must be realized in datacenter room details view to improve its usability:

- [x] Remove the cards view in favor of the table view only
- [x] Show equipments tags in table
- [x] Split table by racks sections with [grouped rows](https://tailwindui.com/components/application-ui/lists/tables#component-3fc4f58b97988d00656bc99fbd1a81ae), sorted by names. Equipments in racks sections must then be ordered by position in the racks from top to bottom.
- [x] Improve readability of equipment table by reducing its width.
- [x] When the infrastructure image is maximized, it should be able to close the modal with escape button or by clicking on the sides.

### PR DESCRIPTION
**Pull request by florianLSP**
fix #41

### CODE REVIEW

File: racksdb/web/ui/src/components/InfrastructureTable.vue [hunk 1]
```diff
@@ -10,41 +10,112 @@ import type {
...
+  const index = rackNames.value.indexOf(rackName)
+  if (index !== -1) {
+    rackDisplayStates.value[index] = !rackDisplayStates.value[index]
   }
 }
 
-function allInfrastructurDetails() {
-  showRack.value = false
-  showInfrastructureRacks.value = true
+function sortRooms() {
```
Comment (rezib) [inline racksdb/web/ui/src/components/InfrastructureTable.vue:57]: This function does not sort rooms, please fix its name.

File: racksdb/web/ui/src/components/InfrastructureTable.vue [hunk 2]
```diff
@@ -10,41 +10,112 @@ import type {
...
+        return { ...{ equipmentType: 'misc' }, ...misc } as Misc
+      })
+    )
+  })
+  // Sort equipment by slot in descending order
+  return equipments.sort((a, b) => b.slot - a.slot)
 }
 
-function popUp(name: string, equipment: equipmentType) {
+function popUp(name: string, equipmentType: 'nodes' | 'network' | 'storage' | 'misc') {
```
Comment (rezib) [inline racksdb/web/ui/src/components/InfrastructureTable.vue:111]: Why don't you use `EquipmentType` type defined above?

File: racksdb/web/ui/src/components/InfrastructureTable.vue [hunk 3]
```diff
@@ -54,320 +125,126 @@ function popUp(name: string, equipment: equipmentType) {
...
-          <td class="px-6 py-4">{{ network.rack }}</td>
-          <td @click="popUp(network.name, 'network')" class="px-6 py-4 capitalize cursor-pointer">
-            {{ network.type.id }}
-          </td>
-        </tr>
-      </template>
+      <tbody>
+        <template v-for="(rack, index) in rackNames" :key="rack">
+          <tr class="border-b border-t bg-gray-50 h-14 text-lg">
+            <th>
```
Comment (rezib) [inline racksdb/web/ui/src/components/InfrastructureTable.vue:174]: Please use colspan instead of empty cells on these rows.

**Review by rezib — CHANGES_REQUESTED** [review]
Some nodes are not listed in the correct order in my test:

image

In this example, `mecn0058` shoud appear below `mecn0059`.

The whole equipment type cell is clickable:

image

This is a bit weird as there is no separation from the cell on its left or right. I would set the link on the text instead, with a different style and hover change so the users can quickly figure out it can be clicked. 

The equipment type popup should rely on headless UI Modal for more accessibility (eg. esc key or click away to close modal).

There is also an issue with developed racks panels and inverted sorting order. For example, consider this situation where _rackE01_ is developed:

image

After I click the button to invert sorting order, _rackE01_ is closed and _R5-A02_ is open:

image

I would expect _rackE01_ to stay open instead.

File: racksdb/web/ui/src/components/InfrastructureTable.vue (hunk 2 above)
Comment (florianLSP) [reply racksdb/web/ui/src/components/InfrastructureTable.vue:111]: I did not think about using `EquipmentType`, but I made the change and used it.

File: racksdb/web/ui/src/components/InfrastructurePopUp.vue [hunk 4]
```diff
@@ -1,98 +1,41 @@
-<!--Copyright (c) 2022-2023 Rackslab
```
Comment (rezib) [inline racksdb/web/ui/src/components/InfrastructurePopUp.vue:1]: You introduce the notion of _PopUp_ for a component generally named _modal_ (or _dialog_) by the community, can you please rename it to use the same name as the community?

File: racksdb/web/ui/src/components/InfrastructurePopUp.vue [hunk 5]
```diff
@@ -1,98 +1,41 @@
...
-  id: String,
-  infrastructure: {
-    type: Object as PropType<Infrastructure>,
+import { Dialog, DialogPanel } from '@headlessui/vue'
+
+defineProps({
+  showPopUp: Boolean,
+  popUpContent: {
+    type: Object as PropType<
+      NodeEquipment | NetworkEquipment | StorageEquipment | MiscEquipment | undefined
```
Comment (rezib) [inline racksdb/web/ui/src/components/InfrastructurePopUp.vue:21]: The complete equipment object is not needed in this component, only its type is required. You should better change this to restrict the manipulation on the type object only.

File: racksdb/web/ui/src/components/InfrastructureTable.vue [hunk 6]
```diff
@@ -10,364 +10,239 @@ import type {
...
+import { ref, onMounted } from 'vue'
+import {
+  BarsArrowDownIcon,
+  BarsArrowUpIcon,
+  ChevronDownIcon,
+  ChevronUpIcon
+} from '@heroicons/vue/24/outline'
+import InfrastructurePopUp from '@/components/InfrastructurePopUp.vue'
+
+const rackNames: Ref<Array<string>> = ref([])
```
Comment (rezib) [inline racksdb/web/ui/src/components/InfrastructureTable.vue:25]: You use an array `rackNames` whose indexes are used to determine the display boolean of the corresponding rack in `rackDisplayStates` array. You should better use a `Record<string, boolean>` object to associate racks names to their boolean display state.

File: racksdb/web/ui/src/components/InfrastructureTable.vue [hunk 7]
```diff
@@ -10,364 +10,239 @@ import type {
...
   } else {
     var layout = props.infrastructureDetails.layout
 
-    for (let x = 0; x < layout.length; x++) {
-      var result = layout[x][equipment].find((item) => item.name == name)
-      if (result) {
-        popUpContent.value = result
-        break
+    if (equipmentType) {
+      for (let x = 0; x < layout.length; x++) {
```
Comment (rezib) [inline racksdb/web/ui/src/components/InfrastructureTable.vue:116]: You can avoid this equipment search in the infrastructure by directly passing the modal content in parameter of the function, as you already have all the context when you call this function in the template.

**Review by rezib — CHANGES_REQUESTED** [review]
I still have some requests before merging this PR, please have a look at the comments below.

File: racksdb/web/ui/src/components/InfrastructurePopUp.vue (hunk 4 above)
Comment (florianLSP) [reply racksdb/web/ui/src/components/InfrastructurePopUp.vue]: I replaced all popup mentions by modal.

File: racksdb/web/ui/src/components/InfrastructureTable.vue (hunk 6 above)
Comment (florianLSP) [reply racksdb/web/ui/src/components/InfrastructureTable.vue]: I made the changes and remove rackNames, rackDisplayStates and use Reccord.

**Review by rezib — CHANGES_REQUESTED** [review]
Yet some comments, mostly about naming, to polish this thing.

File: racksdb/web/ui/src/components/InfrastructureModal.vue [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+
+        <div class="pt-5">
+          <p class="font-medium">Netifs:</p>
+          <div v-for="netif in modalContent.netifs" :key="netif.type">
+            <p class="pl-5">
+              Type: <span class="italic text-sm">{{ netif.type }}</span>
+            </p>
+            <p class="pl-5">
+              Bandwith:
+              <span class="italic text-sm">{{ (netif.bandwidth * 8) / 1000 ** 3 }}Gb/s</span>
```
[the remark this answers is not in the document]
Comment (florianLSP) [reply racksdb/web/ui/src/components/InfrastructureModal.vue:119]: Okay, I removed the multiplied by 8!

File: racksdb/web/ui/src/components/InfrastructureTable.vue [hunk 9]
```diff
@@ -10,364 +10,232 @@ import type {
...
+      .forEach((rack) => {
+        reversedDisplayRacks[rack] = displayRacks.value[rack]
+      })
+    displayRacks.value = reversedDisplayRacks
   }
+  alphabeticalOrder.value = !alphabeticalOrder.value
+}
+
+function getRacks() {
+  var layout = props.infrastructureDetails.layout
```
[the remark this answers is not in the document]
Comment (florianLSP) [reply racksdb/web/ui/src/components/InfrastructureTable.vue:73]: I removed the function and relocated the code directly in onMounted()

File: racksdb/web/ui/src/components/InfrastructureTable.vue [hunk 10]
```diff
@@ -10,364 +10,232 @@ import type {
...
+      ...part.misc.map((misc) => {
+        return { ...{ equipmentType: 'misc' }, ...misc } as Misc
+      })
+    )
+  })
+  // Sort equipment by slot in descending order
+  return equipments.sort((a, b) => b.position.height - a.position.height)
 }
 
+function modal(
```
[the remark this answers is not in the document]
Comment (florianLSP) [reply racksdb/web/ui/src/components/InfrastructureTable.vue:115]: I made the change!

File: racksdb/web/ui/src/components/InfrastructureTable.vue [hunk 11]
```diff
@@ -10,364 +10,232 @@ import type {
...
+  displayRacks.value[rackName] = !displayRacks.value[rackName]
 }
 
-function popUp(name: string, equipment: equipmentType) {
-  if (showPopUp.value) {
-    showPopUp.value = !showPopUp.value
-  } else {
-    var layout = props.infrastructureDetails.layout
+function sortRacks() {
+  if (displayRacks.value) {
```
[the remark this answers is not in the document]
Comment (florianLSP) [reply racksdb/web/ui/src/components/InfrastructureTable.vue:59]: No it cannot be false, so I removed the condition.

File: racksdb/web/ui/src/components/InfrastructureTable.vue [hunk 12]
```diff
@@ -10,364 +10,232 @@ import type {
...
+function displayRackEquipment(rackName: string) {
+  displayRacks.value[rackName] = !displayRacks.value[rackName]
 }
 
-function popUp(name: string, equipment: equipmentType) {
-  if (showPopUp.value) {
-    showPopUp.value = !showPopUp.value
-  } else {
-    var layout = props.infrastructureDetails.layout
+function sortRacks() {
```
[the remark this answers is not in the document]
Comment (florianLSP) [reply racksdb/web/ui/src/components/InfrastructureTable.vue:58]: To give it a correct name, I called it invertRacksSort().

File: racksdb/web/ui/src/components/InfrastructureModal.vue [hunk 13]
```diff
@@ -27,75 +31,71 @@ defineProps({
...
+      <div v-if="'cpu' in modalContent">
         <p class="font-semibold flex justify-center text-purple-700 text-xl capitalize py-5">
-          {{ popUpContent.type.id }}
+          {{ modalContent.id }}
         </p>
 
         <div class="pt-5">
           <p class="pl-5">
-            Model: <span class="italic text-sm">{{ popUpContent.type.model }}</span>
+            Model: <span class="italic text-sm">{{ modalContent.model }} Bonsoir</span>
```
[the remark this answers is not in the document]
Comment (florianLSP) [reply racksdb/web/ui/src/components/InfrastructureModal.vue:50]: Oups, I removed it :)

**Review by rezib — APPROVED** [review]

**Pull request merged**